File: backend/app/schemas/diagrams.py
"""Pydantic schemas for architecture diagram generation API."""

from __future__ import annotations

from pydantic import BaseModel, Field


class VisNodeResponse(BaseModel):
    id: str
    label: str
    kind: str
    group: str = ""
    description: str = ""
    file_path: str = ""
    x: float = 0.0
    y: float = 0.0


class VisEdgeResponse(BaseModel):
    id: str
    source: str
    target: str
    label: str = ""
    kind: str = "default"


class GraphViewResponse(BaseModel):
    nodes: list[VisNodeResponse] = Field(default_factory=list)
    edges: list[VisEdgeResponse] = Field(default_factory=list)


class MermaidDiagramsResponse(BaseModel):
    flowchart: str
    sequence: str
    class_diagram: str


class DiagramBundleResponse(BaseModel):
    repo_id: str
    mermaid: MermaidDiagramsResponse
    system_architecture: GraphViewResponse
    dependency_graph: GraphViewResponse
    markdown: str
    stats: dict[str, int | str] = Field(default_factory=dict)


class DiagramExportResponse(BaseModel):
    repo_id: str
    format: str
    content: str

File: frontend/src/api/diagrams.ts
import { apiClient } from "@/api/client";

export interface VisNode {
  id: string;
  label: string;
  kind: string;
  group: string;
  description: string;
  file_path: string;
  x: number;
  y: number;
}

export interface VisEdge {
  id: string;
  source: string;
  target: string;
  label: string;
  kind: string;
}

export interface GraphView {
  nodes: VisNode[];
  edges: VisEdge[];
}

export interface MermaidDiagrams {
  flowchart: string;
  sequence: string;
  class_diagram: string;
}

export interface DiagramBundle {
  repo_id: string;
  mermaid: MermaidDiagrams;
  system_architecture: GraphView;
  dependency_graph: GraphView;
  markdown: string;
  stats: Record<string, string | number>;
}

export async function fetchDiagrams(encodedRepoId: string): Promise<DiagramBundle> {
  const { data } = await apiClient.get<DiagramBundle>(`/diagrams/${encodedRepoId}`, {
    timeout: 60_000,
  });
  return data;
}

export async function exportMarkdown(encodedRepoId: string): Promise<string> {
  const { data } = await apiClient.get<string>(`/diagrams/${encodedRepoId}/export/markdown`, {
    timeout: 60_000,
    responseType: "text",
    transformResponse: [(v) => v],
  });
  return data;
}
